docs/ORTModule_ModuleWithLoss_Wrapper.md

```diff
@@ -0,0 +1,114 @@
+
+def create_model_with_loss(self):
+    model_with_loss = ModuleWithLoss(self.model, self.args, self.label_smoother)
+    model_with_loss.compute_model_plus_loss_internal = types.MethodType(Trainer.compute_loss, model_with_loss)
+
+    return model_with_loss
+
+def _set_signature_columns_if_needed(self):
+    if self._signature_columns is None:
+        # Inspect model forward signature to keep only the arguments it accepts.
+        import inspect
+
+        if isinstance(self.model, ModuleWithLoss):
+            signature = inspect.signature(self.model._original_model.forward)
+        else:
+            signature = inspect.signature(self.model.forward)
+
+        self._signature_columns = list(signature.parameters.keys())
+        # Labels may be named label or label_ids, the default data collator handles that.
+        self._signature_columns += list(set(["label", "label_ids"] + self.label_names))
+
+def compute_loss(self, model_with_loss, inputs, return_outputs=False):
+    # Run model forward + loss compute.
+    if isinstance(self.model, ModuleWithLoss):
+        # ORTModule Does not support the BatchEncoding Type so we have to convert to a dict.
+        dict_inputs = dict(inputs.items())
+        return model_with_loss(dict_inputs, return_outputs)
+    else:
+        return super().compute_loss(model_with_loss, inputs, return_outputs)
+
+```
+
+## Creating the ModuleWithLoss Wrapper
+
+### 1. Define `ModuleWithLoss` Class
+
+- Create a class named `ModuleWithLoss` that extends `nn.Module`.
+- Implement the __init__ method to initialize the wrapper with the original model, training arguments (`args`), and label smoother.
+- Add an instance variable to store the original model (`_original_model`).
+- Include the label smoothing functionality if required.
+
+### 2. Override the `forward` Method
+
+- In the `ModuleWithLoss` class, override the forward method.
+- Within this method, call `compute_model_plus_loss_internal` to compute the loss using the original model, inputs, and return_outputs.
+- Return the computed loss.
+
+### 3. Define the `module` and `config` properties
+
+- Implement the `module` property to provide access to methods and properties on the original model.
+- Implement the `config` property to return the configuration of the original model.
+
+## Initializing the Wrapper
+
+### 4. Create a function to initialize the wrapper:
+- Implement a function, such as `create_model_with_loss()`, to create an instance of `ModuleWithLoss` using the original model, arguments, and label smoother.
+- Assign `compute_model_plus_loss_internal` as a method of `Trainer.compute_loss` to the wrapper instance.
+- Return the created wrapper instance.
+
+### 5. Override the necessary functions:
+- Override the `_set_signature_columns_if_needed` function if needed.
+- Update the function to handle the case where the model is an instance of `ModuleWithLoss`.
+
+### 6. Override the `compute_loss` function if needed:
```

Comment:
the steps are specialized, I don't think for other training script, users can follow those steps  for warpping the loss.